### ISSUE TITLE
3ds models reading is broken

### ISSUE THREAD

**Issue by 375gnu** [CelestiaProject/Celestia#1343]
**Describe the bug**
For any 3ds model from the current content repository i see errors "Failed to read 3DS chunk" and nothing is shown.

**Comment by ajtribick** [comment]
Works fine for me on current master (cbdd9de6) with CelestiaProject/CelestiaContent@cc0bc0b. (Windows 10, Ubuntu 21.10, tested with Cassini, Huygens, Galileo, Apollo 11)

**Comment by 375gnu** [comment]
I will check

**Comment by 375gnu** [comment]
not related to locale. here is a log file
[log.txt](https://github.com/CelestiaProject/Celestia/files/7978380/log.txt)
.

**Comment by ajtribick** [comment]
Did a quick check, Mir also works fine for me. Will take another look later.

**Comment by ajtribick** [comment]
Only thing that comes to mind is that we don't have a .gitattributes file in CelestiaContent. Maybe your Git has decided for some reason that .3ds files are text files and is doing line-ending conversion?

First 16 bytes of mir.3ds should be:
```
4D 4D C1 99 07 00 02 00 0A 00 00 00 03 00 00 00
```

if you've got that 0A converted to 0D 0A then that might be causing the problems.

**Comment by 375gnu** [comment]
not my case as i'm on unix

**Comment by ajtribick** [comment]
Well, here's the mir.3ds file I have on my system (currently working), can you check with this?
[mir.zip](https://github.com/CelestiaProject/Celestia/files/7987169/mir.zip)

**Comment by ajtribick** [comment]
@375gnu are you still seeing this issue, including with the attached file? I've added logging and I get the same trace, except that `in` status is good. And the test is running fine in the automated build.

**Comment by 375gnu** [comment]
@ajtribick your file is 100% identical to mine

**Comment by ajtribick** [comment]
Ok. Well, since I can't reproduce it on any of my environments, or with the CI, I think you're going to have to debug this one.

**Comment by dave-kaye** [comment]
Hello 375gnu - thank you for fixing the GTK planet rings rendering bug so quickly.

Running Linux openSUSE version Leap 15.2.  For the 3ds models of MIR, Apollo 11, Galileo and Cassini none of them will render in present master in either GTK or QT front ends.

Once again, I took my 1.6.2 GTK compiled front end only and activated it (all other Celestia files are from the compiled master tree) and all of the 3ds models rendered successfully.

Took a bit editor to the MIR 3ds file and did not see the 0D 0A linefeed, only the 0A bit.  That is not the problem.

Got all the 2001 models from the Celestia Motherlode.  Any 3ds model will not render in master branch; checked several with the 1.6.2 GTK front end and they ran successfully.  Problem is not with the Celestia github data download site.

Hope this helps you narrow down your search.

p.s. Please do not solve problem by dropping 3ds support.  I downloaded over 100 3ds models from the Motherlode site, trying to cmod them and then test them will be.....

**Comment by dave-kaye** [comment]
Continuing to look at the 3ds model issue.  If the code works on one Linux distribution and not another, then it probably is in base programs.  In this case it would include C++, openGL and the newly added eigen.

I am running eigen3 version 3.3.4.  Ubuntu 21.10 is running eigen3 version 3.3.9.  So I found an upgrade for my system to eigen 3.4.0.  It did not help the problem.  However, in the install file was the comment:
--------------------------------------------------------
# The OpenGL support test fails
%bcond_with opengl_test

# Tests fail for different reasons within the test-suite itself; disable for now
# See e.g. https://gitlab.com/libeigen/eigen/-/issues/2088, https://gitlab.com/libeigen/eigen/-/issues/2092
------------------------------------------------------
Don't know if this has any bearing on the problem.

I started downloading a series of Celestia git versions, compiling them and testing them.  The test models were from standard extras: Mir, Cassini and Apollo 11; and from 2001: Discovery, Monolith (TMA-2), Orion III spaceplane and Space Station V.  A result of "works" means all test models rendered successfully.  A result of "fails" means all models failed to render except Orion III and Cassini (with date change).  In timing order the results are:

06-27-21  git 05cb719  works
10-26-21  git ec232b7  works
11-05-21a  git ef10800  works
11-05-21b  git 71bfec7  works
11-05-21c  git f3e5444 fails
11-06-21  git 80d098f  fails
11-30-21  git 41936fc   fails
12-30-21  git 85e5bae  fails
02-11-22  git 6b856b3  fails
04-17-22  git 4c0fa12  fails

I jumped back and forth to narrow in the git version, but have listed them all here.  This is valid for my system and build only.

I hope this helps.

**Comment by dave-kaye** [comment]
I have continued to look at the 3ds problem.  I laboriously took the 3dsread.cpp file from the working 11-05-21b commit and slowly added a few lines from the failing 11-05-21c commit, compiled and tested the models.  No rendering problems were encountered until the very end.  The code was approximately line 540:
```
M3DScene* Read3DSFile(std::istream& in)
{
    std::uint16_t chunkType;
    if (!readUshort(in, chunkType) || chunkType != M3DCHUNK_MAGIC)
    {
        fmt::print(std::clog, "Read3DSFile: Wrong magic number in header\n");
        return nullptr;
    }

    std::int32_t chunkSize;
    if (!readInt(in, chunkSize) || chunkSize < 6)
    {
        fmt::print(std::clog, "Read3DSFile: Error reading 3DS file top level chunk size\n");
        return nullptr;
    }

    fmt::print(std::clog, "3DS file, {} bytes\n", chunkSize + 6);

    auto scene = std::make_unique<M3DScene>();
    std::int32_t contentSize = chunkSize - 6;

    std::int32_t bytesRead = read3DSChunks(in, contentSize, processTopLevelChunk, scene.get());
    /*if (bytesRead < 0) { return nullptr; }
    if (bytesRead != contentSize)
    {
        return nullptr;
    }*/

    return scene.release();
}
```
In the code was a check for bytes read.  The non-rendering models were returning bytes read of -1, the read failure flag.  Up until this commit, the return value was never checked.  I commented out the lines as seen above, and the models rendered successfully.  As a fix, similar lines of code exist in the same routine on the present git master.  Commenting them should allow the models to render successfully.

I do not know if this fix is system dependent or not, as some are not seeing the problems.

I started back tracking the issue.  I got it back to read3DSChunks routine.  It went further back to read3DSChunk routine.  In this function I added a line of code similar to your logger function in the default case, but it is in the unknown chunk types case looking at the chunkType variable.  
```
switch (processedSize)
    {
    case READ_FAILURE:
        return READ_FAILURE;
    case UNKNOWN_CHUNK:
        fmt::print(std::cout, "unknown Chunk type {:04x}\n", chunkType);
        in.ignore(contentSize);
        return in.good() ? chunkSize : READ_FAILURE;
    default:
        if (processedSize != contentSize)
        {
            fmt::print(std::cout, "Chunk type {:04x}, expected {} bytes but read {}\n", chunkType, contentSize, processedSize);
            return READ_FAILURE;
        }
        return chunkSize;
    }
```
Then I ran the problem models.  The results are:

MIR unknown chunk type b000
Space Station V unknown chunk type b000
Discovery unknown chunk type b000
Monolith unknown chunk type b000
Apollo 11 chunk type 3d3d expected bytes 1481504 read bytes -9

I don't think I can do much more.  It may be that there is an unaccounted for chunk type of b000 or some issue in the model file.  The Apollo 11 model appears to be a file issue.  I do not know enough about your code and 3ds models to be of further assistance.

As Celestia does not crash with the problem model files, and historically does not appear to have issues with the files, I am just going to patch my system to comment out the offending lines.  I will wait for a more permanent fix from you.

**Comment by 375gnu** [comment]
@dave-kaye good investigation!

**Comment by dave-kaye** [comment]
I have put the jumper comment section in my copy of the master code.  The models now render, but a new bug has surfaced.  The Discovery model is rendering with sections missing.

I did my compile over time thing again.  It is rendering successfully in 11-30-21 commit and not completely in 12-30-21 commit.

See the ajtribick commit, but I don't know where it is going since it appears to be outside of the Celestia tree and I don't know where/which to pick up and see if I can test.

**Comment by munix9** [comment]
@dave-kaye could you please test the [AppImage](https://github.com/CelestiaProject/Celestia/issues/333) to see if the problem occurs there as well?
It is based on openSUSE Leap 15.3 (FYI: 15.2 is EOL).

**Comment by dave-kaye** [comment]
Right now I don't have the capability to move from Leap 15.2.   Not for a long time it appears. 

I have a trick that I am using that may help ajtribick see the problems.  I am installing the OS onto a USB 3.0 flash drive, then my boot order is DVD>USB>HardDrive.  Pop the flash drive into the port, boot and I am in that operating system.

**Comment by dave-kaye** [comment]
I followed the link #1363  to a github repository ajtribick/Celestia where it appears the rewritten code exists.  There is a pull request in that repository for changes in 3ds codes.  I downloaded the complete repository code as a file celestia-rewrite3ds.zip.

I extracted the cel3ds directory files and put them in my copy of the Celestia 1.7.0 master.  Then I recompiled and tested the Qt based executable.

All test models were from standard extras: Mir, Cassini and Apollo 11; and from 2001: Discovery, Monolith (TMA-2), Orion III spaceplane and Space Station V.  All models loaded and rendered successfully.  The new problem of the Discovery model not having some sections rendering did NOT show up in this test.

I didn't see anything in a quick inspection of the models that seemed to be out of the ordinary.  Right now, I don't see any bugs using this code on my system.

**Comment by dave-kaye** [comment]
@munix9 since the pending code push from ajtribick appears to solve the problem, I am going to wait until it is done before doing anything.  It is just waiting on review/approval to go the the master branch.  The present appImage will be history at that time.  If there are problems on the push, I can look at the appImage then, but I have never messed with an appImage so I can't guarantee I will be able to activate it.

**Comment by munix9** [comment]
no problem, but it's trivial to use appimages.
short form: download + make executable + start
eg.
```
$ wget -O celestia-1.7.0-git-x86_64.AppImage https://download.opensuse.org/repositories/home:/munix9:/unstable/AppImage/celestia-latest-x86_64.AppImage
$ chmod 755 celestia-1.7.0-git-x86_64.AppImage
$ ./celestia-1.7.0-git-x86_64.AppImage
```
It would be interesting to know if it works for you.

**Comment by dave-kaye** [comment]
Tried the appimage file.  Did not run.  Fails with multiple missing files - major ones are libstdc++.so.6, libm.so.6 and libc.so.6.  Appears to need the c libraries that match the system that it is prepared for in order to open and boot the program.

**Comment by munix9** [comment]
> Tried the appimage file. Did not run. Fails with multiple missing files - major ones are libstdc++.so.6, libm.so.6 and libc.so.6. Appears to need the c libraries that match the system that it is prepared for in order to open and boot the program.
> 

ok, thanks for testing.

**Comment by dave-kaye** [comment]
SORRY!!!!!!!!!!!!!

I was just rechecking the Celestia update in the pull request.  It does NOT solve the model rendering issues.  The compiled pull request was loaded into the main directory structure and not the usr directory structure.  So what I tested was my last compiled known good version of the Celestia history.

The pull request version is failing to render the test models in the same manner as previously described.

To many copies of Celestia.  Very sorry for the confusion....

**Comment by ajtribick** [comment]
@dave-kaye - unfortunately there is very little I can do with that statement. Please provide the 3ds file that's failing and the log output of when it is loaded (e.g. by passing `--log output.txt` on the command line). Plus specify compiler version used, OS, and architecture.

**Comment by ajtribick** [comment]
I have now also built Celestia on an OpenSUSE Leap 15.3 Live USB. No issues loading 3ds files with that one either (either the existing master branch or my new branch).

**Comment by ajtribick** [comment]
It's also working with the AppImage on OpenSUSE 15.3.

Steps I am using to test:

1. Set year to 1987
2. Go to "Mir"
3. Verify the model is loaded
4. Set date to Dec 5th 1995
5. Go to "Galileo Probe" (note that the main Galileo model is a cmod, so not relevant here)
6. Verify the model is loaded

**Comment by ajtribick** [comment]
I've updated the PR branch to tolerate an EOF status after skipping a chunk (or the trailing bytes of a chunk), which is my best guess as to what the issue might be. An EOF *should not* happen if the reader doesn't try to read beyond the end of a well-formed 3ds file. But yeah, once again I give you the standard developer mantra: "it works on my machine".

**Comment by dave-kaye** [comment]
Sorry to burst your mantra, but your supposition that an EOF will not happen appears to be wrong.

For this posting, the following convention will be used:
- phase 0 is the present github Celestia master branch
- phase 1 is the ajtribick Celestia clone without the EOF code and the directory cel3ds copied into the present Celestia github master
- phase 2 is the ajtribick Celestia clone with the EOF code and the directory cel3ds copied into the present Celestia github master

Results: models MIR, Apollo 11 and Discovery (sourced and as listed in previous posts) fail to load and render in phase 0 and phase 1 codes.  All models render successfully in phase 2 codes.  Went back and tested the entire group of models that I have been using in previous posts - they all render successfully in phase 2 code.

System used is openSUSE Leap 15.2 on x86-64.  The output of cmake in preparing to compile lists the libraries used.
[cmake output.txt](https://github.com/CelestiaProject/Celestia/files/8599643/cmake.output.txt)

The Celestia log section for the model loading:
[output phase 0.txt](https://github.com/CelestiaProject/Celestia/files/8599651/output.phase.0.txt)
[output phase 1.txt](https://github.com/CelestiaProject/Celestia/files/8599652/output.phase.1.txt)
[output phase 2.txt](https://github.com/CelestiaProject/Celestia/files/8599654/output.phase.2.txt)

The phase 0/1 output appears to match the patterns from the tests of the 11-05-21 b/c Celestia github versions that were reported several days ago.  They also match what I have been seeing in prior versions.  The results keep skipping between specific headers not recognized and a reading failure report.  The header not recognized is b000 for the problem models, except Apollo 11 which is a data read error of a known chunk.

Phase 2 EOF code shows the models loading successfully.  Don't know why an EOF would be generated into the middle of a file read.  The b000 chunk is listed as a frame and is in your enumerated list, but I don't see any code for it.  That could be the root source of the issue and the EOF code is handling it.  I would think that if EOF is embedded into the model, only part of it would copy and a lot of other issues would be seen.  Unless a compiler bug is inserting it into reads in some circumstances.  Apollo appears to be an outlier and could be a bad 3ds file with some little tweak point in the model mesh wrong.  I see a lot of model creators complaining about hidden mesh sections and other issues when they are trying to get their model running.  Maybe one of the model developers on the forum would see something.

Then again, who am I to argue with success?

The models used are:
[mir.zip](https://github.com/CelestiaProject/Celestia/files/8599702/mir.zip)
[Discovery.zip](https://github.com/CelestiaProject/Celestia/files/8599704/Discovery.zip)
[apollo.zip](https://github.com/CelestiaProject/Celestia/files/8599705/apollo.zip)

Thank you for continuing to look at this issue and all the effort you are giving to Celestia.  I will await the official code push/release.

p.s. is 375gnu having the same issues or just reporting on other's problems?  Does this fix it?

**Comment by ajtribick** [comment]
Aha thanks for the confirmation... this libstdc++ bug appears to be exactly the behaviour that's occurring on failures here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96161

Which explains why the EOF fix works.

**Comment by dave-kaye** [comment]
OK, I may not argue with success but I will look a gift horse in the mouth...

I was curious as to the EOF, so I used the log to trace it back to the skipChunk routine.  Some of the prior versions of the IF statements were complex and I was wondering if it was EOF or poor c++ parsing of the IF.  I rewrote the code to expand the logging level:
```
bool skipChunk(std::istream& in, M3DChunkType chunkType, std::int32_t contentSize)
{
    GetLogger()->debug("Skipping {} bytes of unknown/unexpected chunk type {}\n", contentSize, chunkType);
    in.ignore(contentSize);
    if (in.eof())
    {
        GetLogger()->error("skipChunk:eof - Error skipping {} bytes of unknown/unexpected chunk type {}\n", contentSize, chunkType);  
        return true;
    }
    if (in.good())
    { 
         GetLogger()->error("skipChunk:good - Error skipping {} bytes of unknown/unexpected chunk type {}\n", contentSize, chunkType);   
         return true;
    }
    GetLogger()->error("skipChunk - Error skipping {} bytes of unknown/unexpected chunk type {}\n", contentSize, chunkType);
    return false;
}
```
The results for the 3 models are here:
[output.txt](https://github.com/CelestiaProject/Celestia/files/8603060/output.txt)

The results for the 3 models show a large number of calls to the skipChunk routine, but the final call is always through the EOF section.  And only a single, final call is through EOF.  I ran a working model Orion III and it too showed a large number of calls to skipChunk, but no final call to the EOF section (results not attached).

Still curious, I went looking at the EOF.  There is no EOF character in Linux, but the history of computing has control-z  (hex 1A) as a true EOF character.  It is a legend that this still shows up in some Windows areas.  The 3ds file specification requires the texture names to be in the old DOS specification for naming.  I can verify this because I have had to go in with a hex editor to work on texture names when capitalization went out of sync between *.3ds and filenames.  I used a hex editor to scan the three model files and found a lot of control-z characters embedded in the files, but the last character was not control-z.  Finally, I added a single hex 00 character to the end of the model files.  The log files show the last call to skipChunk is from the same chunk style and number of bytes, but it goes through the "good" section of IF and not the EOF section.

OK, I think I have beaten this poor horse to death on this issue.

Thanks for all your efforts.

-- signing off --
